Use a dedicated typeclass to guide coproduct induction

joroKr21 opened the pull request:
This disentangles the `Coproduct` induction logic in `ToResponse`
derivation from the actual encoding of values.

The net result is greatly reduced compilation times for large
`Coproducts`.

Fixes #923

joroKr21 commented:
I will think about adding a test

vkostyukov commented:
Very interesting. How much of a compile time improvement we're talking about?

joroKr21 commented:
WIP for now as I want to see if shapeless has a builtin combinator we can use without sacrificing performance. Maybe something like `LiftAll` and `Unifier`? or `LeftFolder` / `RightFolder`.

In one case I cared to measure the compile time improved from ~4min to ~6sec.

vkostyukov commented:
Nice! Thanks for the extra details @joroKr21! Yeah, I'd be interested in seeing if Shapeless has anything we'd be able to reuse. Also, I'm not minding this thin type-class at all. Just let me know when you think it's ready for review.

joroKr21 commented:
It seems like `Folder` is exactly what we want, but the phantom `ContentType` parameter makes it difficult to express.

vkostyukov commented:
I'd personally be fine with this new type-class. Let's just make sure we get more people to take a look.

vkostyukov commented on core/src/main/scala/io/finch/ToResponse.scala:
```diff
@@ -124,14 +124,29 @@ object ToResponse extends HighPriorityToResponseInstances {
+trait CoproductToResponse[C <: Coproduct] extends ToResponse[C]
+object CoproductToResponse {
+  type Aux[C <: Coproduct, CT] = CoproductToResponse[C] { type ContentType = CT }
+  def apply[C <: Coproduct](implicit ctp: CoproductToResponse[C]): Aux[C, ctp.ContentType] = ctp
```
Should we drop this `apply` method? I don't see it's being used anywhere?

joroKr21 replied:
Just habit, I will remove it.

vkostyukov commented on core/src/main/scala/io/finch/ToResponse.scala:
```diff
@@ -124,14 +124,29 @@ object ToResponse extends HighPriorityToResponseInstances {
+  ): Aux[C, CT] = ctr
+}
+
+trait CoproductToResponse[C <: Coproduct] extends ToResponse[C]
```
I'd add an extra new line here.

vkostyukov commented on core/src/main/scala/io/finch/ToResponse.scala:
```diff
@@ -124,14 +124,29 @@ object ToResponse extends HighPriorityToResponseInstances {
+  ): Aux[C, CT] = ctr
+}
+
+trait CoproductToResponse[C <: Coproduct] extends ToResponse[C]
```
Just an idea. How do you feel about moving this new typeclass under the `ToResponse` companion object? This way we can reduce the top-level/user-facing API surface. I'd imagine something like `ToResponse.FromCoproduct` should do it:

```scala
object ToReponse {
  type Aux = ...
  trait FromCoprodct extends ToResponse[...]
}

```

joroKr21 replied:
Fine by me, I was wondering the same thing :)

vkostyukov replied:
Unless you already working on it, I'd be happy to see this merged in its current state and then addressing this comment later. Whatever you prefer - just le me know.

ImLiar commented:
LGTM  👍

vkostyukov commented:
FWIW, I tried compiling `examples` project with it - no noticeable impact. Perhaps, the examples are too simple for this to "kick in".

joroKr21 commented:
I addressed the comments. @vkostyukov as you noticed there is no easy way to test this.

I published locally the changes here and noticed a tiny slowdown of about 10sec compared to the workaround I have in our codebase, but it's still a huge improvement wrt the baseline. The difference in the workaround is a specialized `toCoproductService` that takes an implicit `ToResponse.FromCoproduct` instead of the more general `ToResponse`.

Anyway, I think it's fine like this.

vkostyukov commented:
Love it! Thank you @joroKr21!
